### DISCUSSION TITLE
repeat_effect_until with a once called initialisation

### DISCUSSION THREAD

**Discussion by mathisloge** [NVIDIA/stdexec#1434] [Q&A] [answered]
Sorry, I really try to understand the API, but it is sometimes quite hard with the sparse examples (or do I look at the wrong sources?) 

I'm trying to achive a repeated process, but with an initial set of async operations which shouldn't be repeated.
I always end up with either the initial block being executed with each repetition or the block to be repeated not being executed at all.
I'm stuck at `how to call repeated_work now?` below: 

```cpp
auto repeated_work = then([](auto&& context) {
     // do some work
     return value;
   })
| then([](auto&& value){ 
    // do smth with the value
     return value; 
})
| then([](auto&& value){ return value.valid() })
| exec::repeat_effect_until();

schedule(scheduler)
 | then([](){ // do some long running work }
 | then([](auto&&... values){ return context; }
 // init finished.
 // how to call repeated_work now? 

```

**Comment by maikel** [comment, marked as answer]
Try `let_value`. If your context type is move-constructible then this will be good enough. Otherwise, you need some helper senders. 

Take care when using `exec::repeat_effect_until` with the pipe operator. It can be surprising what will be repeated. 

```cpp
#include <stdexec/execution.hpp>
#include <exec/single_thread_context.hpp>
#include <exec/repeat_effect_until.hpp>

#include <print>

class MyContext {
public:
    explicit MyContext(int data): data_{data} {}

    auto receiveMessage() {
        return stdexec::just(data_);
    }

private:
    int data_;
};

int main()
{
    exec::single_thread_context context;
    auto scheduler = context.get_scheduler();

    auto sndr = //
        stdexec::when_all(
            stdexec::then(stdexec::schedule(scheduler), [] { return MyContext{42}; }),
            stdexec::just(1))
        | stdexec::let_value([](MyContext& context, int& x) {
            return //
                context.receiveMessage()
                | stdexec::then([&](int value) { 
                      std::print("{}: {}\n", x, value);
                      return x++ == 3;
                  })
                | exec::repeat_effect_until();
        });

    stdexec::sync_wait(sndr);
}
```
https://godbolt.org/z/3YMEKnPEd

**Reply by mathisloge** [reply]
Ahh, I've missed the when_all and let_value functor params as ref. 
At least on gcc14 the template errors that are generated for the pipe operators aren't quite helpful with what is expected. 

But thanks for the fast help and your answers! :)
